### ISSUE TITLE
Bug: Generating names based on custom training data doesn’t work

### ISSUE THREAD

**Issue by huftis** [SidharthMacherla/conjurer#22]
Generating names with `buildNames()` results in an error message if one specifies a *custom* data frame of names. Here’s a reprex:

```r
library(conjurer)
df_names = data.frame(names = c("Oliver", "Jack", "Harry"),
                      stringsAsFactors = FALSE)
new_names = buildNames(df_names, numOfNames = 3, minLength = 5, maxLength = 7)
#> Error in unlist(alphaList, use.names = FALSE) : 
#> object 'alphaList' not found
```

<details>

R version 3.6.3 (2020-02-29)
Platform: x86_64-suse-linux-gnu (64-bit)
Running under: openSUSE Tumbleweed

Matrix products: default
BLAS:   /usr/lib64/R/lib/libRblas.so
LAPACK: /usr/lib64/R/lib/libRlapack.so

locale:
[1] C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] conjurer_1.1.0

loaded via a namespace (and not attached):
[1] compiler_3.6.3
</details>

**Comment by SidharthMacherla** [comment]
Hi @huftis, thanks for calling this out along with a helpful example. Here is my preliminary evaluation and next steps.
**Background**: A matrix(technically a table) of alphabets is generated out of the training data and is further used to generate the names.
**Current problem(most likely)**: There are only three names with about 12 alphabets. This means the matrix as explained above will be a sparse matrix and does not have enough data to generate the next alphabet. This is most likely the cause of this issue.
**Next step for myself**: I will attempt to see what best alternative could help in generating a name no matter how sparse the matrix is. One idea that comes to my mind is going down Laplace smoothing type solution where all values are incremented by 1 to make the matrix dense.
Based on my thoughts and experiments, I will respond. I will post an update on my progress next week. 
_Note: Additionally, a better error handling message is needed at the least._

**Comment by SidharthMacherla** [comment]
**Issue fixed**. Submitted new version 1.1.1 on CRAN. This issue will be formally closed once the new version is published on CRAN.

**Detailed explanation for author's record**:
- Preliminary evaluation of the issue is not correct.
- **Problem**: Since the training data is very less, the matrix generated from the training data is less. This means that the nextAlphaProb function could generate an alphabet that is not present in the alphaMatrix(neither first nor all). This leads to error when the alphaMatrix(first or all) tries to get subset by that _nextAlpha_.
- **Solution**: In order to solve this, the following steps are taken.
  - In _buildNames_ function, the _topAlphas_ is set to 13 or the number of training names provided. Whichever is less.
  -  In _buildNames_ function, _nextAlphaProb_ is first checked with _placement_ "all". If there is no such alphabet combination, then it is run using the _placement_ "first". If even that does not exist, then the name is stopped there.
  - This means that the names could be smaller than the requested length. Additionally, the names will not sound (phonetically) similar to the training data. This is particularly because the _nextAlphaProb_ is using _alphaMatrixFirst_ instead of _alphaMatrixAll_. 
  - Since user is responsible for providing sufficient training data, a note is printed when the length of _topAlphas_ is less than 13.

**Comment by huftis** [comment]
Thanks. I’ve tested the v1.1.1 branch, and it seems to work (when there are at least 13 names).

BTW, I’m not sure your above analysis is entirely correct. I have just tested the old version v1.1.0 with a data frame with several hundred names, and I still got the same error message. But anyway, it’s fixed in v1.1.1.

**Comment by SidharthMacherla** [comment]
New version 1.1.1 released and available on CRAN now. Following test case given by @huftis is resolved. Additionally, 2 more tests using library(babynames) are also given below. 

**Test 1** (_as given in the bug report_)
```r
install.packages("conjurer")
library(conjurer)
set.seed(1)
df_names = data.frame(names = c("Oliver", "Jack", "Harry"),stringsAsFactors = FALSE)
new_names = buildNames(df_names, numOfNames = 3, minLength = 5, maxLength = 7)
print(new_names)
```
**Test 1 Output** 
```r
Warning message:
In buildNames(df_names, numOfNames = 3, minLength = 5, maxLength = 7) :
  Training data is not large enough. Expect less than minimum length names and/or names that do not seem like  training data
> print(new_names)
[1] "hl" "jl" "jl"
```

**Test 2** (_Test with just 3 names for training_ )
```r
install.packages("babynames")
library(babynames)
set.seed(100)
df_names <- as.data.frame(babynames$name[1:5])
new_names = buildNames(df_names, numOfNames = 3, minLength = 5, maxLength = 7)
print(new_names)
```
**Test 2 Output**
```r
Warning message:
In buildNames(df_names, numOfNames = 3, minLength = 5, maxLength = 7) :
  Training data is not large enough. Expect less than minimum length names and/or names that do not seem like  training data
> print(new_names)
[1] "ela"  "elb"  "mabb"
```

**Test 3** (_Test with 5000 names for training_ )
```r
set.seed(101)
df_names <- as.data.frame(babynames$name[1:5000])
new_names = buildNames(df_names, numOfNames = 3, minLength = 5, maxLength = 7)
print(new_names)
```
**Test 3 Output**
```r
[1] "cland"  "anchra" "henath"
```
**Conclusion:** 
As the training data increases, the names generated seem phonetically more similar to the training names.
